core/src/main/scala/org/http4s/headers/Forwarded.scala

```diff
@@ -0,0 +1,178 @@
+ */
+
+package org.http4s.headers
+
+import java.net.{Inet4Address, Inet6Address}
+
+import cats.data.NonEmptyList
+import cats.syntax.either._
+import org.http4s._
+import org.http4s.util.{Renderable, Writer}
+
+object Forwarded
+    extends HeaderKey.Internal[Forwarded]
+    with HeaderKey.Recurring
+    with ForwardedRenderers
+    with parser.ForwardedModelParsing {
+
+  final case class Node(nodeName: Node.Name, nodePort: Option[Node.Port] = None)
+
+  protected[http4s] val NodeNameIpv4 = Node.Name.Ipv4
+
+  object Node {
+    def apply(nodeName: Name, nodePort: Port): Node = apply(nodeName, Some(nodePort))
+
+    sealed trait Name { _: Product => }
+
+    object Name {
+      case class Ipv4(address: Uri.Ipv4Address) extends Name
+      case class Ipv6(address: Uri.Ipv6Address) extends Name
+      case object Unknown extends Name
+
+      def apply(address: Uri.Ipv4Address): Name = Name.Ipv4(address)
+      def apply(address: Inet4Address): Name = apply(Uri.Ipv4Address.fromInet4Address(address))
+      def apply(a: Byte, b: Byte, c: Byte, d: Byte): Name = apply(Uri.Ipv4Address(a, b, c, d))
+
+      def apply(address: Uri.Ipv6Address): Name = Name.Ipv6(address)
+      def apply(address: Inet6Address): Name = apply(Uri.Ipv6Address.fromInet6Address(address))
+      def apply(a: Short, b: Short, c: Short, d: Short, e: Short, f: Short, g: Short, h: Short)
+          : Name = apply(Uri.Ipv6Address(a, b, c, d, e, f, g, h))
+    }
+
+    sealed trait Port { _: Product => }
+
+    object Port {
+      private[this] final case class C(value: Int) extends Port {
+        override def productPrefix: String = "Port"
+      }
+
+      def fromInt(num: Int): ParseResult[Port] =
+        checkPortNum(num).toLeft(C(num))
+
+      def unapply(port: Port): Option[Int] =
+        PartialFunction.condOpt(port) { case C(num) => num }
+    }
+
+    /** Opaque type for obfuscated identifiers.
+      *
+      * @param value obfuscated identifier with leading '_' (underscore) symbol.
+      *
+      * @see [[https://tools.ietf.org/html/rfc7239#section-6.3]]
+      */
+    sealed abstract case class Obfuscated private (value: String) extends Name with Port
+
+    object Obfuscated {
+      def fromString(s: String): ParseResult[Obfuscated] =
+        new ModelNodeObfuscatedParser(s).parse
+
+      /** Unsafe constructor for internal use only. */
+      private[http4s] def apply(s: String): Obfuscated = new Obfuscated(s) {}
+    }
+
+    def fromString(s: String): ParseResult[Node] = new ModelNodeParser(s).parse
+  }
+
+  sealed abstract case class Host private (host: Uri.Host, port: Option[Int])
+
+  object Host {
+    private[this] def apply(host: Uri.Host, port: Option[Int]) = new Host(host, port) {}
+
+    def apply(uriHost: Uri.Host): Host = apply(uriHost, None)
+
+    def from(uriHost: Uri.Host, port: Int): ParseResult[Host] =
+      checkPortNum(port).toLeft(apply(uriHost, Some(port)))
+
+    def from(uriHost: Uri.Host, port: Option[Int]): ParseResult[Host] =
```

Comment:
What's the use case for one that takes an `Option`?  This seems most useful if you have a `Uri.Authority`.

Reply:
Well, it is a bit useful in tests where `Gen[Option[Int]]` is used to generate `Host`-s with or without port numbers. But I'm inclining to agree that this override seems redundant.

Reply:
... Also it turned out that it is used in parsers :) So some internals leverage this form of `from` a little bit.